models/train.py

```diff
@@ -0,0 +1,236 @@
+#! /usr/bin/env python3
+from emoe_v1 import ElasticMoELlamaForCausalLM
+from emoe_config import ElasticMoELlamaConfig
+from emoe_utils import profile, writing
+from packed_dataset import PackedDataset
+
+import torch
+from torch.optim.lr_scheduler import LambdaLR
+from torch.utils.data import DataLoader, IterableDataset
+from torch.utils.tensorboard import SummaryWriter
+
+# from datasets import load_dataset
+
+import os
+import glob
+from pickle import dump
+import cProfile
+import multiprocessing
+from jsonargparse import CLI
+
+
+PROC_COUNT = multiprocessing.cpu_count()
+
+def train(
+        # Trace
+        profiling_path:str="profile", 
+        tensorboard_path:str="tensorboard",
+        logging_step_interval:int=1,
+
+        # Schedule
+        epochs:int=1,
+        warmup_steps:int=0,
+        max_steps:int=100,
+        eval_step_interval:int=1000,
+        checkpoint_path:str="checkpoint",
+        checkpoint_step_interval:int=3000,
+
+        # Data
+        data_cache_dir:str="data",
+        block_size:int=10,
+        micro_batch_size:int=1,
+        batch_size:int=1,
+        chunks=1,
+        num_workers:int=4,
+
+        # Optimizer
+        learning_rate:float=1e-4,
+        beta1:float=0.9,
+        beta2:float=0.95,
+        weight_decay:float=0.01,
+        decay_learning_rate:bool=True,
+        gradient_clip:float=1.0,
+):
+    """
+    Train a model using the Elastic MoE Llama model.
+    """
+
+    train_path_str = os.path.join(data_cache_dir, "red_pajama_sample",  "train/*")
+    eval_path_str = os.path.join(data_cache_dir, "red_pajama_sample",  "validation/*")
```

Comment:
can we download the datasets if they dont exist? For us this will be easy, but if we show the code to others that will be nice to have.